torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,120 @@
+import collections
+import warnings
+from functools import partial
+
+import torch
+from torch.testing._internal.common_cuda import (TEST_CUDA)
+import torch.testing._internal.common_methods_invocations as cmi
+from torch.testing._core import _dispatch_dtypes
+from torch.testing import (all_types_and_complex_and,
+                           all_types_and_complex,
+                           all_types_and_half,
+                           all_types,
+                           complex_types,
+                           floating_and_complex_types,
+                           floating_types_and_half,
+                           floating_types,
+                           integral_types,
+                           floating_types_and,
+                           floating_and_complex_types_and,
+                           integral_types_and,
+                           all_types_and,
+                           )
+
+COMPLETE_DTYPES_DISPATCH = (
+    all_types,
+    all_types_and_complex,
+    all_types_and_half,
+    floating_types,
+    floating_and_complex_types,
+    floating_types_and_half,
+    integral_types,
+    complex_types,
+)
+
+EXTENSIBLE_DTYPE_DISPATCH = (
+    all_types_and_complex_and,
+    floating_types_and,
+    floating_and_complex_types_and,
+    integral_types_and,
+    all_types_and,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
+
+# Class to tag the dynamically generated types.
+
+
+class _dynamic_dispatch_dtypes(_dispatch_dtypes):
+    pass
+
+
+def get_supported_dtypes(op, sample_inputs_fn, device):
+    # Returns the supported dtypes for the given operator and device pair.
+    assert device in ['cpu', 'cuda']
+    if not TEST_CUDA and device == 'cuda':
+        warnings.warn("WARNING: CUDA is not available, information pertaining to CUDA dtypes could be wrong")
+
+    supported_dtypes = set()
+    for dtype in all_types_and_complex_and(torch.bool, torch.bfloat16, torch.half):
+        samples = sample_inputs_fn(op, device, dtype, False)
+        for sample in samples:
+            try:
+                op(sample.input, *sample.args, **sample.kwargs)
+                supported_dtypes.add(dtype)
+            except RuntimeError as re:
+                if "not implemented for" not in str(re):
+                    raise re
+                pass
+
+    return _dynamic_dispatch_dtypes(supported_dtypes)
+
+
+def dtypes_dispatch_hint(dtypes):
```

Comment:
Comment for this function explaining it and its intended use

I'm still curious how this handles the _AND variants vs the dtype functions without them?

Reply:
Could you give an example of the case you are concerned about?

Reply:
Let's say the set of dtypes is float, double, and complex64. How does this function produce floating_types_and(complex64) and not floating_types(complex64)?

Reply:
Unfortunately, for the above case,
```python
_, name = dtypes_dispatch_hint((torch.float64, torch.float, torch.complex64))
print(name)
```

it prints,
```
floating_and_complex_types_and()
```
which is the closest. But I hope none of the operator supports only `torch.complex64` but not `torch.complex32`

So the idea is if we get a perfect match for the dispatch function without `_and`, we return that.

https://github.com/pytorch/pytorch/blob/6a5f23b6d7cb1f15371c96e846b3630fecba571c/torch/testing/_internal/opinfo_helper.py#L24-L33

Else we return the closest and version and the operators which weren''t present in the and version.

https://github.com/pytorch/pytorch/blob/6a5f23b6d7cb1f15371c96e846b3630fecba571c/torch/testing/_internal/opinfo_helper.py#L35-L41

Reply:
But isn't `floating_and_complex_types_and()` incorrect? Because that would also include complex128, and we should only return `and` if we're actually using it, too.

That's why I was suggesting we find the exact match or, failing that, the set with the most matches (and no extra elements)

Reply:
Ah. Now I get it. Will update it. Thanks!